Incorporate getBlockLabel into useBlockDisplayInformation

david-szabo97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->
Related https://github.com/WordPress/gutenberg/issues/27876
## Description
We introduced a function for blocks that returns a dynamic label based on the block's attributes: `__experimentalLabel`. ([Example](https://github.com/WordPress/gutenberg/blob/eeea281928071a348336fe4d875d79e7008e69e9/packages/block-library/src/template-part/index.js))

The Block inspector didn't take this into account.  To make sure this function is taken into account throughout the codebase, this PR incorporates it into the already existing and widely used `useBlockDisplayInformation` hook. This also simplified the code in `BlockTitle`.

## How has this been tested?
1. Open block editor
2. Select a `Template Part` block
3. Open block inspector
4. Make sure the inspector shows the name of the template part
5. Smoke test everything else

## Screenshots <!-- if applicable -->

(Block inspector says "Header" instead of "Template Part")
![image](https://user-images.githubusercontent.com/2256104/107976243-896a2580-6fb9-11eb-877c-eb56e776c4bf.png)

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Incorporate getLabel into useBlockDisplayInformation

david-szabo97 committed:
Update readme

david-szabo97 committed:
Update test

jeyip commented:
## Testing
### Behavior
- [x] Block inspector displays name of template part
- [x] When updating template part name, block inspector displays updated name 
- [x] When creating a new template part, block inspector displays name of new template part
- [x] When inserting template parts in other templates, block inspector still displays name of template part  

### Browsers
- [x] Chrome
- [x] Edge
- [x] Firefox
- [x] Safari
- [ ] ~IE11~

**Note:**
Didn't test IE11 because it renders a white screen of death https://github.com/WordPress/gutenberg/issues/28249

> Smoke test everything else
@david-szabo97 Is there anything else you had in mind to smoke test?

jeyip reviewed (commented):
Would it be possible to add a small e2e test to verify that the block inspector displays the template part name?

jeyip commented on packages/block-editor/README.md:
```diff
@@ -574,14 +574,16 @@ _Related_
+Hook used appropriate information for display reasons.
+It takes variations and custom `label` function into account.
+
+In order to try to find a variation match we need to things:
```
> In order to try to find a variation match we need to things:

In order to try to find a variation match we need **two** things:

jeyip commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -1,11 +1,16 @@
 
-1. Block's client id to extract its current attributes.
-2. A block variation has `isActive` prop defined with a matcher function.
+In order to try to find a variation match we need to things:
```
> In order to try to find a variation match we need to things:

In order to try to find a variation match we need **two** things:

jeyip commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -1,11 +1,16 @@
 # useBlockDisplayInformation
 
-A React Hook that tries to find a matching block variation and returns the appropriate information for display reasons. In order to try to find a match we need two things:
+Hook used appropriate information for display reasons.
```
> Hook used appropriate information for display reasons.

Not sure how best to revise this. Did you mean "Hook used to return appropriate information for display reasons" ?

david-szabo97 replied:
Duh, looks like I'm unusable after 6 PM. Fixed them, thanks!

jeyip commented on packages/block-editor/src/components/use-block-display-information/README.md:
```diff
@@ -1,11 +1,16 @@
-1. Block's client id to extract its current attributes.
-2. A block variation has `isActive` prop defined with a matcher function.
+In order to try to find a variation match we need to things:
+1\. Block's client id to extract it's current attributes.
```
> Block's client id to extract it's current attributes.

`NB:` Block's client id to extract **its** current attributes.

jeyip commented on packages/block-editor/src/components/use-block-display-information/index.js:
```diff
@@ -22,49 +25,67 @@ import { store as blockEditorStore } from '../../store';
+			const dynamicTitle = getBlockLabel( blockType, attributes );
+			const variationInfo = getVariationInfo( variations, attributes );
+
+			const title =
```
The use of the `getBlockLabel` to generate a "dynamic" title makes sense to me, but, with that being said, I don't have a lot of historical context with `use-block-display-information`.

Also, @david-szabo97, do you happen to know what the exact differences are between a block label and title? The label appears to be a canonical name, while the title is something customized by the user (definitely correct me if I'm wrong).

talldan replied:
> The label appears to be a canonical name, while the title is something customized by the user (definitely correct me if I'm wrong).

I don't think the `title` is really a thing that can be defined anywhere, it seems to be what we use as a variable name when deciding whether to show the block's name (e.g. 'Heading') or the label (the value returned by the `__experimentalLabel` property on a block).

jeyip replied:
Oh I see -- I had it reversed in my head. This makes sense. Thanks @talldan!

jeyip reviewed (commented):
+1 from me when all comments are addressed

gziolo commented:
I want to confirm that the helper introduced in #18132 by @talldan is used as intended. As far as I can tell, `__experimentalGetLabel` was designed to add more flexibility to the block title once inserted in the editor.

Separately from this PR, should we promote this API to the stable name like `getBlockTitle` or whatever fits best? It would be great to add usage details, too.

talldan commented:
Thanks for the review ping. 

From a user experience perspective, I'd be a bit nervous about removing one of the few places we display the actual title of a block (e.g. Heading, Link, Template Part).

> I want to confirm that the helper introduced in #18132 by @talldan is used as intended. As far as I can tell, __experimentalGetLabel was designed to add more flexibility to the block title once inserted in the editor.

> Separately from this PR, should we promote this API to the stable name like getBlockTitle or whatever fits best? It would be great to add usage details, too.

I think it needs some work before becoming stable. I had a PR open to tackle some improvements - https://github.com/WordPress/gutenberg/pull/19664 - but there was so much disagreement that it never really progressed.

One of the good bits of feedback though was making `__experimentalGetLabel` a selector instead of an API function in the blocks package. I think that makes a lot of sense, we don't want to add unnecessary stuff to that API.

jameskoster reviewed (approved):
It works :)

> From a user experience perspective, I'd be a bit nervous about removing one of the few places we display the actual title of a block (e.g. Heading, Link, Template Part).

I think this will be improved (for template parts at least) when we implement the changes suggested in #26599

david-szabo97 committed:
Fix documentation

david-szabo97 committed:
Add e2e test

david-szabo97 committed:
Revert activateTheme

david-szabo97 commented:
@jeyip 
> In your testing notes @david-szabo97, you said to "Smoke test everything else". I think we should also test functionality that > implements useBlockDisplayInformation. Is this what you meant? If that's the case, then we should probably verify proper  functionality for:
> BlockNavigationBlockSelectButton
> BlockSwitcherDropdownMenu

Correct! Anything that references `useBlockDisplayInformation`.

Added an e2e test and updated the comments.

Copons commented:
I don't have much to say about the code, but I wonder if this might make the UX worse. 🤔 

As far as I can see, right now there are 2 core blocks making use of the `__experimentalLabel` (excluding those that do it only in accessibility context, which I'm not sure what it entails): Template Part and Navigation Link.

Navigation Link can afford to use a dynamic title as its descriptions are... descriptive enough (e.g. "Add a link to a URL").

Template Part's description, on the other hand, kinda rely on the "Template Part" title to convey a complete meaning: "Edit the different global regions of your site, like the header, footer, sidebar, or create your own".

Doesn't this read awkward?

<img width="271" alt="Screenshot 2021-02-17 at 14 26 30" src="https://user-images.githubusercontent.com/2070010/108218782-e89c7700-712c-11eb-8d4a-611f1b7b2206.png">

Would it make sense to have Template Part's label return something like `Template Part: Header`?
Or maybe it could be better to fine tune the description to make it more generic.

(To be fair, though, I wonder if me clinging to the block title is a power user/geek bias, while most people wouldn't care at all.)

mtias commented:
I'd rather leave the block inspector untouched here (reflecting the default block name). The idea for displaying "Header" there will be to create a block variation of the template part block set to the semantic taxonomy of `header`. By being its own variation, it will use the block name, icon, and description best fitting for "Header".

jameskoster commented:
I think that colorising template parts, and using the new icons will help (26599). 

It may also be helpful to revise the existing description, so that it explicitly mentions template parts. Maybe something like:

<img width="363" alt="Screenshot 2021-02-17 at 16 14 26" src="https://user-images.githubusercontent.com/846565/108233022-46d05680-713b-11eb-9652-1aa4550830fd.png">

We should probably explore that in a separate issue though.

ntsekouras commented:
>I'd rather leave the block inspector untouched here (reflecting the default block name). The idea for displaying "Header" there will be to create a block variation of the template part block set to the semantic taxonomy of header. By being its own variation, it will use the block name, icon, and description best fitting for "Header".

I echo @mtias on this. I think this seems block variations territory.

david-szabo97 commented:
Basically, a template part can have any name. So I don't think variations is a good fit here. Like, we can't create a variation for each template part, that doesn't sound good to me 🤔

ntsekouras commented:
>Basically, a template part can have any name. So I don't think variations is a good fit here. Like, we can't create a variation for each template part, that doesn't sound good to me 🤔

Block Card seems more `block type` information and that's why it feels better to me to leave it as is. The name currently is shown in the Inspector controls and here on Block Card too, with a rather irrelevant description/icon.

<img width="306" alt="Screenshot 2021-02-19 at 9 54 55 AM" src="https://user-images.githubusercontent.com/16275880/108475193-10a8e900-7299-11eb-8ed8-4f70122bd1d4.png">

While it makes sense to show the `label` in `Navigation List` or `breadcrumbs` that are more user facing and help the user distinguish different template parts in the content, in Block Card the information are for the selected block only.

I'm not sure exactly how we can handle this as variations but is worth exploring. What do you think?

vindl commented:
> Basically, a template part can have any name. So I don't think variations is a good fit here. Like, we can't create a variation for each template part, that doesn't sound good to me 🤔

We can create one for the supported categories and everything else outside of that can be a generic template part?

If you check out the parent issue in the context of the [milestone](https://github.com/WordPress/gutenberg/issues/24818) it building up on previous item related to block variations:
> - Define header, footer, and sidebar terms (maybe also register them as variations of template part for easier control flow and hide the variations in the inserter.)
> - Apply specific icons to those variations. Color template parts and selection state in the block tree and icon in block description. #27876

mtias commented:
> We can create one for the supported categories and everything else outside of that can be a generic template part?

Yes, correct. Generic template parts should just read as "template part" in the inspector. Same with reusable blocks. We use the "label name" in other contexts like List View, Toolbar, Breadcrumb.

david-szabo97 commented:
Superseded by https://github.com/WordPress/gutenberg/pull/29122
